[jundatan] iP

jundatan opened the pull request:
# Duke
Duke helps you to save and check your tasks anytime you want!

> "The number one app that everyone needs" - somebody

- CLI commands
- Easy to use
- ~~Quick~~ Quickest way to save your tasks

What you need to do is,

1. download it from [here](https://github.com/jundatan/ip/releases/tag/A-Jar)
2. Open it.
3. Add your tasks (Todo, Deadlines, Events)
4. Let it do your job of remembering 👍 

It is absolutely **Freeee**!

Features:
- [x] Managing tasks
- [ ] Reminders
- [ ] AI implemented

#
If you are a java programmer, you can use it to practice Java too. Here's the ``` main ``` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

jundatan committed:
Added Greet, Echo, Exit features.

jundatan committed:
Added Add to List and List feature.

jundatan committed:
Added Mark as Done/Not Done feature.

jundatan committed:
Added Todos, Events Deadlines features.

jundatan committed:
Added Automated Text UI Testing.

jundatan committed:
Added Handling of Errors.

jundatan committed:
Added Delete feature.

jundatan committed:
Added Enums for Commands.

jundatan committed:
Add Save feature

Duke is running on data that is not saved and resets every time the
Duke program is rerun.

Tasks should be saved and used for future rerun of Duke for the users.

Let's add a folder with a text file and call Scanner class to read the
text file and FileWriter class to write in the text file.

Using Scanner class to read the text file and FileWriter class to write
 in the text file helps to store task in a text file as expected.

jundatan committed:
Add dates formatting for deadline

The toString method in Deadline class prints the deadline in a String
format provided by the user.

The toString method should print in the mmm-dd-yyyy format instead of a
string.

Let's import the LocalDate class and parse in the date to the var by.
Use the var by of type LocalDate to print the date in mmm-dd-yyyy format
.

LocalDate class helps to provide a format that can be printed out for
the user.

jundatan committed:
Merge branch 'branch-Level-7'

jundatan committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java

jundatan committed:
Add more classes for OOP

Program is cluttered in static main with little OOP.

Program code looks untidy with no OOP in mind.

Let's add more OOP classes such as Parser to deal with making sense of
the user command, Storage to deal with loading tasks from the file and
saving tasks in the file, TaskList to contain the task list and UI to
deal with user interactions.

These classes can help to further categorize the different objects in
an OOP way which makes the code more readable.

jundatan committed:
Divide classes into packages

Classes are all inside java folder with no clear packaging.

Packaging of classes offer better classification of classes which makes
file management neater and clearer.

Let's group most of the task in duke folder while grouping tasks in
task folder.

Task class and its subclasses should be in a different package from the
rest of the duke classes.

jundatan committed:
Resolve packaging issues

Packages and import having issues but still able to run on IDE.

Let's fix the packages and other related issues for declaration of
packages and import.

Problem with syntax. Only way to fix this.

jundatan committed:
Merge remote-tracking branch 'origin/add-gradle-support'

jundatan committed:
Update build.gradle file path

Unable to run the application.

Cannot run the application due to file path being directed to the wrong
path.

Let's change the file path for application to the correct path.

Only way to run the application is to change the file path to the
correct path.

jundatan committed:
Add JUnit tests for classes in task package and move some classes into
extra folder

Classes are not in the right folder. No JUnit tests currently available.

JUnit tests are required to test the different classes. Classes are not
in right folder will affect the test.

Let's create a seedu folder for easier navigation and create JUnit tests
classes for the classes in the task package.

JUnit tests allow semi-automation of test for the respective classes.

jundatan committed:
Add javadoc for all classes/methods

Minimal javadoc in the some classes.

Easier to read the uses of methods and classes.

Let's add comments to the methods and classes.

Only way to make a javadoc for methods and classes.

jundatan committed:
Follow coding standard

Some parts of the code is not following the coding standard.

Following the coding standard improve the readability of the code.

Let's follow the coding standard by doing proper indentations etc.

Best way to follow the coding standard is to follow it.

jundatan committed:
Add Find feature

No find feature for the user to find what tasks is in the TaskList.

User must be able to find their tasks if it is done or not.

Let's add some find methods in TaskList and Ui to support this issue.

TaskList helps to find the matching tasks and Ui helps to display the
tasks and messages.

jundatan committed:
Merge branch 'branch-A-JavaDoc'

jundatan committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/seedu/duke/Duke.java

jundatan committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/seedu/duke/Duke.java

jundatan committed:
Add checkstyle

There is no checkstyle available in the program.

Require checkstyle to follow the coding standard.

Let's add checkstyle.xml and suppression.xml and use gradle for
checkstyle checking.

One of the way to check the checkstyle is through gradle.

jundatan committed:
Update code according to checkstyle

Code is not passing checkstyle.

Code not following coding standard. Hence, changes are needed to follow
the coding standard.

Let's change up our code according to the checkstyle.

Following checkstyle to correct any coding standard violations.

huangzz125 commented on src/main/java/seedu/duke/Duke.java:
```diff
@@ -0,0 +1,119 @@
+    /**
+     * Represents the set of commands by the user.
+     */
+    public enum Commands {
```
Good use of enums!

huangzz125 commented on src/main/java/seedu/duke/Duke.java:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    /**
+     * Constructor for the Duke class.
+     *
+     * @param filepath The file path for the data stored in the txt file.
```
According to the Checkstyle that we are using, I think there is no extra line between "Constructor ..." and "@param ...". Please do check on this.

jundatan replied:
I checked the coding standard textbook. There is a spacing between the method description and the @params. May i know where you find it on the Checkstyle that says no extra line?

huangzz125 replied:
My bad, you are right. I have been following the warnings produced by the IntelliJ Checkstyle using the checkstyle.xml that says not to leave that empty line.

huangzz125 commented on src/main/java/seedu/duke/Storage.java:
```diff
@@ -0,0 +1,138 @@
+                fw.write("D | " + mark + " | " + element.getDescription() + " | " + ((Deadline) element).getBy());
+            }
+            if (element instanceof Event) {
+                fw.write("E | " + mark + " | " + element.getDescription() + " | "
+                        + ((Event) element).getFrom() + " | " + ((Event) element).getTo());
```
Since you are appending to a string multiple times, I propose the use of String.format() method to make your code look neater.

jundatan replied:
Thank you for the advice! 😄

huangzz125 commented on src/main/java/seedu/duke/TaskList.java:
```diff
@@ -0,0 +1,198 @@
+public class TaskList {
+
+    protected List<Task> tasksList;
+    protected int counter;
```
Suggestion: give this a more specific name, ie taskCounter.

jundatan replied:
Ok, i will make these changes. 👍

huangzz125 reviewed (commented):
Overall, the code is neat, and you have named variables and methods adequately. There are also some use of enums, switch cases,  extensive JavaDoc and more to improve the readability of your code. Well done!

jundatan committed:
Add GUI to the program

The program is using the terminal to print out the outputs.

Printing of the outputs does not look like a chatbot. Using an
application will make it look like a chatbot.

Let's add the classes for GUI and fxml files to use to build the app.

More efficient to use fxml file. Main class and Launcher class for easy
differentiation.

jundatan committed:
Merge branch 'branch-Level-10'

jundatan committed:
Add var-args

There is no var-args usage in the current code.

Allow any number of var-args in the method.

Let's add some var-args into the methods if applicable.

Only way to add var-args.

jundatan committed:
Merge branch 'branc-A-Varargs'

jundatan committed:
Add welcome message to initialization

Welcome message after the user types in something.

It is better to have the welcome message and list of commands when the
program begins.

Let's add the welcome message to the initialization with extra methods.

Best way to show the welcome message is during the initialization of the
app.

jundatan committed:
Fix find feature

The current find feature does not allow users to delete, mark and unmark
in the current find list. list command does not show the users the list
of found tasks.

List and any actions other than adding tasks should be working on the
found tasks.

Let's change up the find command to deliver marking, unmarking,
list and delete command functions on the same find list by using static
variables and checking for same object.

This is the more efficient way of fixing find command.

jundatan committed:
Update some methods from Parser class

Current methods are not as useful.

Added more conditions to make the methods more useful.

Let's add an if statement to check if the var-args exceed a certain
amount.

This is the best way to make use of var-args.

jundatan committed:
Fix some checkstyle erorrs

There are some checkstyle errors in the code.

Require fixing to match the coding standard.

Let's fix the checkstyle errors.

Changing the code is the only way to fix it.

jundatan committed:
Fix some errors

There are errors like index out of bound and list or find command should
be use first before any delete, mark or unmark command.

The program is not working as intended and will cause error.

Let's add more methods and variables to check if list or find command is
used before using the other commands. Additional exception checking to
counter out of bound indexing issues.

Adding more exception checks in the TaskList is optimal.

jundatan committed:
Update javadocs

Missing throw comment in the method comments.

Required to fix the javadoc.

Let's update the javadoc.

Updating the javadoc is the only way.

jundatan committed:
Add assertions

There are no assertions in the code.

Assertions provide error handling on programmer mistakes in the code.

Let's add assertions for the filepath.

It is one of the ways to use assertions effectively to check the file
path.

jundatan committed:
Update on code quality

The duke code quality is poor as there is no much abstraction and SLAP.

Poor code quality can lead to poor readability.

Let's improve on the code quality.

Best way to improve code quality is to use SLAP.

jundatan committed:
Add streams to the code

There is currently no stream usage in the code

Better practice for streams.

Let's add streams to the getCommand method.

One way to use streams effectively.

jundatan committed:
Merge pull request #1 from jundatan/branch-A-Assertions

Add assertions

jundatan committed:
Merge pull request #2 from jundatan/branch-A-CodeQuality

Update on code quality

jundatan committed:
Merge pull request #3 from jundatan/branch-A-Streams

Add streams to the code

jundatan committed:
Fix some checkstyle issues

There are some checkstyle issues.

Required to fix to achieve no coding standard violations.

Let's fix all the checkstyle errors.

Only way to fix the checkstyle errors is to follow coding standards.

jundatan committed:
Add github actions

There is currently no github actions available.

Useful to track pull requests.

Let's create the folder and add gradle.yml

Best way to add github actions.

jundatan committed:
Add tagging features and handle some errors

There are currently no tagging features and some errors.

Fix errors to prevent application from failing. Adding tagging features
allow users to tag their tasks.

Let's handle more errors and allow user to tag their tasks by adding
a list of tags in task class.

Best way to add tags is to put a list of tags in the task class for each
tag object.

jundatan committed:
Merge branch 'branch-C-Tagging'

jundatan committed:
Update gradle.yml

Gradle.yml last three parts causing problem.

Remove the last three parts to prevent the problem.

Let's remove the last three parts of the file.

Only way to remove the issue for now.

jundatan committed:
Update GUI

The current GUI has no application header and icon.

The GUI is currently ugly with the default Java icon and header.

Let's add the GUI header and icon by using scene Image and some String
name called Duke in the Main class start method.

Best way to change the Icon and header of the application to look better
.

jundatan committed:
Update build.gradle to target launcher for shadowjar

The current build.gradle is targetted towards duke class.

Required to build jar file successfully.

Let's change the application build.gradle destination to the launcher
instead.

Best and easiest way to build a working shadowjar.

jundatan committed:
Update on minor coding standard violation

Switch case has no egyptian type brackets.

Add egyptian type brackets to follow coding standards.

Let's add some egyptian type brackets.

Only way to follow the coding standards.

jundatan committed:
Improve code quality and fix a bug

There are still some room of improvement for the code quality. There is
a bug where the tag will duplicate.

The bug can cause major issues.

Let's add some more inline methods and remove the bug by resetting the
tag string for each element in the writing method for the storage.

Best way to fix the current bug.

jundatan committed:
Update on code quality from the bot

Some code quality issues with the code.

There is a need to improve on code quality to make it more readable.

Let's fix up the remaining code quality.

Only way to fix the code quality is to follow the standards of code
quality.

jundatan committed:
Improve GUI

GUI currently has no color background, text font and height is too high.

GUI require some improvements.

Let's change the background color by adding color to our dialogboxes and
add setfont to the dialogbox. Let's also adjust the height of the
dialogbox.

Setting the background color and the font. Adjusting the height in
screenbuilder is one of the better ways to improve gui.

jundatan committed:
Add tag description error

There is no tag description error.

Required to check for empty tag description.

Let's add tag description error for the tag command.

Best way to do this is to throw duke exception when there is an empty
tag description.

jundatan committed:
Fix find feature and change a bit of duke personality

Find feature currently has a bug where it shows the found list even
though it cannot find any matches in the list. Duke talks in the default
way.

Need to fix the find feature so that it will not show the found list
when there is no matches. Change in duke personality as it is quite dull
.

Let's fix the find feature by setting isFind to false when there is no
matches. Let's change some text in the UI to make duke have more
personality.

Best way to fix this is shown as above.

jundatan committed:
Fix checkstyle error

There is a whitespace before an 'if'.

Required to fix checkstyle error.

Let's fix the checkstyle error.

Only way to fix it is to put a whitespace.

jundatan committed:
Fix delete command

Current stage does not request for a new list after deletion.

Important to ask for a new list after some deletions as indexing will
change.

Let's fix the delete command by adding a condition if it isFind to
detect which lists to output and change the isDisplayList boolean to
false after finishing what it has to be done.

Best way to do it is the way shown above.

jundatan committed:
Add user guide, fix tasks adding bug and add Ui picture

Currently, there is no User guide and adding tasks will not ask for
display of new list. There is no screenshot of the Ui in the docs too.

Need to fix all these for better user use.

Let's update the user guide, add the Ui screenshot and fix the bug.

Only way to do it.

jundatan committed:
Update error handling for deadline date format

When deadline date format is wrong, program throw an error that is not
properly handle.

It can cause disruption to the system when it is not handled.

Let's catch the DateTimeParseException and throw a duke exception.

Best way to catch the exception error.

jundatan committed:
Update Ui screenshot name

Ui name is in wrong format.

Need to fix or it cannot be in the showcase.

Let's change the name.

Only way to do it.

jundatan committed:
Update Ui screenshot and text art for the Ui.

Text art is currently buggy looking.

Text are looks bad.

Let's change the text art.

Only way to change it.

jundatan committed:
Update file issues with jar

There is currently an error for txt file for jar as it does not exist.

Creates a big issue as data are not saved.

Let's fix it by making a new directory.

One of the better ways to fix it.

jundatan committed:
Update user guide on user guide and file directory

There is no warning to the user to not delete data directory.

If users do not heed advice to not delete data folder, data will be lost
.

Let's add a warning to the user guide and update on file directory.

Best way to do it is through updates.
